Fixed data dependency to propagate return values of functions to call sites

rajeevgopalakrishna opened the pull request:
This PR addresses https://github.com/crytic/slither/issues/192 to fix data dependency to propagate return values of functions to call sites.

With this fix, for the below contract:
```
contract PropagateThroughReturnValue {
  uint var_dependant;
  uint var_state;

  function foo() public {
    var_dependant = bar();
  }

  function bar() internal returns (uint) {
    return (var_state);
  }
}
```

the `data_dependency` printer output below shows that `var_dependent` in `foo` depends on `var_state` returned from `bar` because of the call site `var_dependant = bar();` in `foo`:

```
Contract PropagateThroughReturnValue
+---------------+---------------+
|    Variable   |  Dependencies |
+---------------+---------------+
| var_dependant | ['var_state'] |
|   var_state   | ['var_state'] |
+---------------+---------------+
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/crytic/slither?pullRequest=193) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/crytic/slither?pullRequest=193) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/crytic/slither?pullRequest=193) it.</sub>

montyly commented on slither/analyses/data_dependency/data_dependency.py:
```diff
@@ -161,11 +162,27 @@ def pprint_dependency(context):
+                        found_return = True
+                        for ir in node.irs_ssa:
+                            if isinstance(ir, Return):
+                                return_values[function] = ir.values
```
We should use a `collections.defaultdict` and the operator `+=`, otherwise function with multiple return statement will not work

rajeevgopalakrishna replied:
Sorry, my bad. Multiple returns is an obvious bug-will fix.

Will also look into memoization of return values.

rajeevgopalakrishna commented:
As suggested above, moved the capturing of return values (with memoization) to properties in `slither/core/declarations/function.py` from `slither/analyses/data_dependency/data_dependency.py`. Fixed the bug where multiple returns within a function would be missed.
